### ISSUE TITLE
placeholder SSN ideas

### ISSUE THREAD

**Issue by JunjieLeiCoe** [JunjieLeiCoe/preprocess#3]
This is the test case; 
```R
# Sample dataframe
df <- data.frame(
  AssociateID = c(1, 2, 2, 3, 4, 5, 5, 5),
  EmployeeSSN = c(NA, "123-45-6789", NA, NA, "987-65-4321", NA, NA, NA),
  stringsAsFactors = FALSE
)
df <- generateSSNPlaceholders(df, "AssociateID", "EmployeeSSN")
print(df)
```

```R 
# Main function in mind; pending implementation to the main package;

generateSSNPlaceholders <- function(data, associate_id_column, ssn_column) {
  associate_ids_with_missing_ssn <- data %>%
    filter(is.na(!!sym(ssn_column))) %>%
    distinct(!!sym(associate_id_column)) %>%
    mutate(placeholder_ssn = sprintf("999-99-%05d", seq_len(n())))

  
  data <- data %>%
    left_join(associate_ids_with_missing_ssn, by = associate_id_column) %>%
    mutate(!!ssn_column := ifelse(is.na(!!sym(ssn_column)), placeholder_ssn, !!sym(ssn_column))) %>%
    select(-placeholder_ssn)

  return(data)
}
```

In the context of the dplyr package in R, the sym function and the !! operator work together to enable non-standard evaluation (NSE), which is a way of programming where expressions are evaluated in a specific context rather than in the standard R manner. This technique is particularly useful when writing functions that manipulate data frames using column names specified as strings or variables.

**Comment by JunjieLeiCoe** [comment]
In the context of the `dplyr` package in R, the `sym` function and the `!!` operator work together to enable non-standard evaluation (NSE), which is a way of programming where expressions are evaluated in a specific context rather than in the standard R manner. This technique is particularly useful when writing functions that manipulate data frames using column names specified as strings or variables.

Let's break down what's happening in this part of the code:

```r
filter(is.na(!!sym(ssn_column))) %>%
    distinct(!!sym(associate_id_column))
```

### The `sym` function

- `sym` is short for "symbol" and is a function in the `rlang` package (which `dplyr` relies on for NSE). `rlang` is a toolbox for programming with the tidyverse.
- `sym()` takes a string as input and turns it into a symbol, which `dplyr` can then use to refer to a column within a data frame.
- For example, if `ssn_column` is `"EmployeeSSN"`, `sym(ssn_column)` converts the string `"EmployeeSSN"` into a symbol that `dplyr` functions can use to refer to the `EmployeeSSN` column in the data frame.

### The `!!` operator (bang-bang)

- The `!!` operator is used to force the evaluation of the expression that follows it. This technique is called "unquoting" in the context of tidy evaluation (part of non-standard evaluation).
- When combined with `sym()`, `!!` tells `dplyr` to use the value of the symbol as a column name in the data frame. This allows you to programmatically specify which column to operate on based on a string variable.
- In essence, `!!sym(ssn_column)` tells `dplyr` to "look at the string value stored in `ssn_column`, convert it to a symbol, and then use that symbol as if it were a direct reference to a column in the data frame."

### Practical Example

Suppose you have a function that accepts a data frame and the name of a column to filter on as a string. Inside this function, you want to use `dplyr` to filter rows based on this column. Using `sym` and `!!` together allows your function to work with the column name specified as a string argument.

This approach is powerful for writing general-purpose data manipulation functions where the specific columns to operate on are not known in advance but are passed to the function as parameters.

In summary, `!!sym(ssn_column)` dynamically refers to a column in a data frame, where the name of the column is stored in the `ssn_column` variable. This technique is particularly useful for writing tidyverse-friendly functions that need to operate on columns specified at runtime.

**Comment by JunjieLeiCoe** [comment]
Certainly! Let's dive into a more detailed example using the `:=` operator for dynamic assignment within a `dplyr` pipeline, especially useful for programmatically naming and mutating columns in a dataframe.

### Scenario:

Imagine you're working with a dataset of sales data, and you want to write a function that dynamically calculates the total sales amount by multiplying the number of units sold (`UnitsSold`) by the unit price (`UnitPrice`). The catch is that both the name of the new column to store the total sales and the names of the columns to use in the calculation are provided as strings at runtime.

Here's how you can accomplish this using `:=` for dynamic assignment:

### Sample Data

First, let's create a sample dataframe:

```r
library(dplyr)

# Sample dataframe
sales_data <- tibble(
  ProductID = c(1, 2, 3),
  UnitsSold = c(10, 15, 20),
  UnitPrice = c(2.50, 5.00, 7.50)
)
```

### Function Definition

Next, define a function that takes the dataframe and column names as arguments and adds a new column with the calculated total sales:

```r
calculate_total_sales <- function(data, units_column, price_column, new_column_name) {
  # Using dynamic column names and assignment with `:=`
  data <- data %>%
    mutate(!!new_column_name := !!sym(units_column) * !!sym(price_column))
  
  return(data)
}
```

### Usage

Now, use this function to calculate the total sales and store it in a new column called `"TotalSales"`:

```r
# Calculate total sales
updated_sales_data <- calculate_total_sales(sales_data, "UnitsSold", "UnitPrice", "TotalSales")

# View the updated dataframe
print(updated_sales_data)
```

### Expected Output

This will produce an output where each product's total sales amount is calculated and stored in the new `TotalSales` column:

```
# A tibble: 3 x 4
  ProductID UnitsSold UnitPrice TotalSales
       <dbl>     <dbl>     <dbl>      <dbl>
1          1        10       2.5       25.0
2          2        15       5.0       75.0
3          3        20       7.5      150.0
```

### Explanation

- `!!sym(units_column)` and `!!sym(price_column)` dynamically convert the column name strings into symbols and then unquote them for evaluation within `mutate()`, effectively referencing the actual columns in the data.
- `!!new_column_name :=` dynamically assigns the calculated values to a new column whose name is determined by the `new_column_name` string variable. The use of `:=` enables you to specify the column name programmatically.

This example illustrates the power and flexibility of `dplyr`'s tidy evaluation system, allowing for dynamic data manipulation that adapts to varying column names and requirements.
